lib3h/src/network_engine.rs

```diff
@@ -0,0 +1,14 @@
+use holochain_lib3h_protocol::{protocol::Lib3hProtocol, Lib3hResult};
+/// Common interface for all types of network modules to be used by the Lib3hWorker
+pub trait NetworkEngine {
+    /// Start network communications
+    fn run(&self) -> Lib3hResult<()>;
+    /// Stop network communications
+    fn stop(&self) -> Lib3hResult<()>;
+    /// Terminate module. Perform some cleanup.
+    fn terminate(&self) -> Lib3hResult<()>;
+    /// Handle some data message sent by the local Client
+    fn serve(&self, data: Lib3hProtocol) -> Lib3hResult<()>;
```

Comment:
I was imagining bubbling this all up one level. That is, I think the api to the networking module might be a little cleaner if we make the net crate in core deal with the worker / channel stuff. We could move this trait into lib3h_protocol and instead of `serve` have something like:

```rust
pub type DidWork = bool;

pub trait NetworkEngine {
    /// post a protocol message from core -> lib3h
    fn post(&mut self, msg: Lib3hProtocol) -> Lib3hResult<()>;

    /// a single iteration of the networking process loop
    /// (should be called frequently to not cpu starve networking)
    /// returns a vector of Lib3hProtocol events core will need to handle
    fn poll(&mut self) -> Lib3hResult<(DidWork, Vec<Lib3hProtocol>)>;
}
```

This is the pattern I arrived at for the WSS transport stuff: https://github.com/holochain/holochain-rust/blob/develop/net/src/ipc/transport/mod.rs

We certainly don't have to go this way if others have other ideas, but this is what I was talking about in terms of pushing the complexity of dealing with asynchronous stuff outside of lib3h.

Reply:
I like that. I'd rename `poll` to `process` though since thats the main intent of calling this function and fits the engine mental model.